nomad/jobs_endpoint.go

```diff
@@ -0,0 +1,297 @@
+			var err error
+			var iter memdb.ResultIterator
+
+			// Get the namespaces the user is allowed to access.
+			allowableNamespaces, err := allowedNSes(aclObj, state, allow)
+			if errors.Is(err, structs.ErrPermissionDenied) {
+				// return empty jobs if token isn't authorized for any
+				// namespace, matching other endpoints
+				reply.Jobs = make([]structs.UIJob, 0)
+				return nil
+			} else if err != nil {
+				return err
+			}
+
+			if prefix := args.QueryOptions.Prefix; prefix != "" {
+				iter, err = state.JobsByIDPrefix(ws, namespace, prefix, sort)
+			} else if namespace != structs.AllNamespacesSentinel {
+				iter, err = state.JobsByNamespace(ws, namespace, sort)
+			} else {
+				iter, err = state.Jobs(ws, sort)
+			}
+			if err != nil {
+				return err
+			}
+
+			tokenizer := paginator.NewStructsTokenizer(
+				iter,
+				paginator.StructsTokenizerOptions{
+					WithNamespace: true,
+					WithID:        true,
+				},
+			)
+			filters := []paginator.Filter{
+				paginator.NamespaceFilter{
+					AllowableNamespaces: allowableNamespaces,
+				},
+				// skip child jobs; we'll look them up later, per parent.
+				paginator.GenericFilter{Allow: func(i interface{}) (bool, error) {
+					job := i.(*structs.Job)
+					return job.ParentID == "", nil
+				}},
+			}
+			// only provide specific jobs if requested.
+			if len(args.Jobs) > 0 {
+				// set per-page to avoid iterating the whole table
+				args.QueryOptions.PerPage = int32(len(args.Jobs))
+				// filter in the requested jobs
+				jobSet := set.From[structs.NamespacedID](args.Jobs)
+				filters = append(filters, paginator.GenericFilter{
+					Allow: func(i interface{}) (bool, error) {
+						job := i.(*structs.Job)
+						return jobSet.Contains(job.NamespacedID()), nil
+					},
+				})
+			}
+
+			jobs := make([]structs.UIJob, 0)
+			newJobs := set.New[structs.NamespacedID](0)
+			pager, err := paginator.NewPaginator(iter, tokenizer, filters, args.QueryOptions,
+				func(raw interface{}) error {
+					job := raw.(*structs.Job)
+
+					// this is where the sausage is made
+					uiJob, highestIndexOnPage, err := UIJobFromJob(ws, state, job, args.SmartOnly)
+					if err != nil {
+						return err
+					}
+
+					jobs = append(jobs, uiJob)
+					newJobs.Insert(job.NamespacedID())
+
+					// by using the highest index we find on any job/alloc/
+					// deployment among the jobs on the page, instead of the
+					// latest index for any particular state table, we can
+					// avoid unblocking the RPC if something changes "off page"
+					if highestIndexOnPage > reply.Index {
+						reply.Index = highestIndexOnPage
+					}
+					return nil
+				})
+			if err != nil {
+				return structs.NewErrRPCCodedf(
+					http.StatusBadRequest, "failed to create result paginator: %v", err)
+			}
+
+			nextToken, err := pager.Page()
+			if err != nil {
+				return structs.NewErrRPCCodedf(
+					http.StatusBadRequest, "failed to read result page: %v", err)
+			}
```

Comment:
Aren't these both internal errors?

Reply:
good catch. this was lifted verbatim from existing jobs endpoint, not sure why it called these "bad request"s, but I can change this'n to `http.StatusInternalServerError` (500)

Reply:
[4447725](https://github.com/hashicorp/nomad/pull/20130/commits/4447725dcaee992fb8660229d0c79c20079f2ea3)